### PR TITLE
docs: Use Read The Docs theme, update to Sphinx v3

### PR DESCRIPTION
**Pull request by WizardCM**
### Description

At its core, this PR:

* Bumps the Sphinx version from 2 to 3
  * Fixes warnings generated by updating Sphinx
* Changes the theme from bizstyle to the more commonly recognised Read the Docs
* Improves navigation by linking to most pages via the main sidebar

Additionally, links to the Build Instructions & Plugin Template are now included on the homepage and main sidebar.

  
image

<details>
  <summary>More screenshots</summary>

image

image

image
</details>

I am very open to feedback on colours, additional styling, alternative fixes to Sphinx 3 incompatibilities and more.

I hope to submit another PR later down the line to better organise/group certain information and more accurately define parameters.

### Motivation and Context

The current Documentation pages proide a lot of useful & important information. However, it can often be hard to navigate or easy to get lost.

Sphinx 3 introduces proper directives for `macro`, `struct`, `union`, `enum`, `enumerator`, and [more](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cpp-domain).

Additionally, this should open the door for deploying versioned documentation in the future, so developers can browse documentation for various OBS versions (such as latest release + latest beta + latest master alongside each other).

### How Has This Been Tested?

Deployed using GitHub Pages to https://wizardcm.com/obs-studio/

### Types of changes

 - Documentation (a change to documentation pages)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -36,10 +36,15 @@ jobs:
           SCOPY="copyright = '([A-Za-z0-9, ]+)'"
           RCOPY="copyright = '2017-$(date +"%Y"), Hugh Bailey'"
           sed -i -E -e "s/${SVER}/${RVER}/g" -e "s/${SREL}/${RREL}/g" -e "s/${SCOPY}/${RCOPY}/g" docs/sphinx/conf.py
-      - uses: ammaraskar/sphinx-action@master
+      - uses: totaldebug/sphinx-publish-action@1.2.0
+        env:
+          ACTIONS_STEP_DEBUG: true
```
Comment (WizardCM) [inline .github/workflows/docs.yml:41]: TODO: Remove

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -36,10 +36,15 @@ jobs:
...
         with:
-          docs-folder: "docs/sphinx/"
-      - uses: actions/upload-artifact@v3
+          sphinx_src: 'docs/sphinx'
+          build_only: True
+          target_branch: 'master'
+          target_path: '../home/_build'
+      - uses: actions/upload-artifact@v1
```
Comment (RytoEX) [inline .github/workflows/docs.yml:47]: Why v1?

File: docs/sphinx/reference-libobs-graphics-math.rst [hunk 3]
```diff
@@ -36,4 +36,4 @@ Helper functions/macros for graphics math.
 .. function:: float rand_float(int positive_only)
 
    Generates a random floating point value (from -1.0f..1.0f, or
-   0.0f..1.0f if *positive_only* is set).
+   0.0f..1.0f if *positive_only* is set).
```
Comment (RytoEX) [inline docs/sphinx/reference-libobs-graphics-math.rst:39]: Did this whitespace get changed? Was that intentional?

File: docs/sphinx/reference-libobs-util-serializers.rst [hunk 4]
```diff
@@ -102,9 +102,11 @@ Provides an output serializer used with dynamic arrays.
 Array Output Serializer Structure (struct array_output_data)
 ------------------------------------------------------------
 
-.. type:: struct array_output_data
+.. struct:: array_output_data
 
-.. member:: DARRAY(uint8_t) array_output_data.bytes
+.. code:: cpp
```
Comment (RytoEX) [inline docs/sphinx/reference-libobs-util-serializers.rst:107]: Isn't this code technically C rather than C++?

**Review by RytoEX — CHANGES_REQUESTED** [review]
Was the removal of `typedef` and `type::` throughout necessary? Just want to confirm.

File: .github/workflows/docs.yml (hunk 2 above)
Comment (WizardCM) [reply .github/workflows/docs.yml]: Unintentional. Fixed.

File: docs/sphinx/reference-libobs-graphics-math.rst (hunk 3 above)
Comment (WizardCM) [reply docs/sphinx/reference-libobs-graphics-math.rst]: Unintentional. Fixed.

File: docs/sphinx/reference-libobs-util-serializers.rst (hunk 4 above)
Comment (WizardCM) [reply docs/sphinx/reference-libobs-util-serializers.rst:107]: Technically, yes, but all other code blocks we have on docs are all `.. code:: cpp` so I did this for consistency.

**Comment by WizardCM** [comment]
> Was the removal of `typedef` and `type::` throughout necessary? Just want to confirm.

Yes - per the commit description.

> Sphinx 3 adds a number of C and C++ friendly directives, making documenting our code much easier.

If I don't remove those additional terms, Sphinx spits out a large amount of warnings.

File: docs/sphinx/reference-libobs-util-serializers.rst (hunk 4 above)
Comment (RytoEX) [reply docs/sphinx/reference-libobs-util-serializers.rst:107]: Fair enough.

**Comment by WizardCM** [comment]
Apologies for the commit spam, the Sphinx action is designed to publishing to a site and not generating an artifact so it had some pathing issues that needed to be cleared up.

This should be ready to go. Note the ZIP size has grown due to the theme's additional fonts (FontAwesome, Lato, and Roboto).

**Comment by RytoEX** [comment]
There appear to still be two warnings:
> WARNING: Invalid configuration value found: 'language = None'. Update your configuration to a valid langauge code. Falling back to 'en' (English).
>
> /github/workspace/docs/sphinx/reference-libobs-graphics-graphics.rst:630: WARNING: duplicate label reference-libobs-graphics-graphics:draw functions, other instance in /github/workspace/docs/sphinx/reference-libobs-graphics-graphics.rst

**Review by RytoEX — APPROVED** [review]
Opened the docs artifact locally and navigated around. Everything worked as expected. The remaining warnings in the generation process have been discussed off-thread and are not blockers.

**Pull request merged**